### PR TITLE
Configuration wrapper object with validation.

### PR DESCRIPTION
**Pull request by friso**
This also removes the configuration based mapping entirely.

Build passes.

### CODE REVIEW

File: src/main/java/io/divolte/server/IncomingRequestProcessingPool.java [hunk 1]
```diff
@@ -49,20 +45,20 @@
     private final Optional<KafkaFlushingPool> kafkaPool;
     private final Optional<HdfsFlushingPool> hdfsPool;
 
-    public IncomingRequestProcessingPool() {
-        this(ConfigFactory.load(), (e,b,r) -> {});
-    }
+//    public IncomingRequestProcessingPool() {
+//        this(ConfigFactory.load(), (e,b,r) -> {});
+//    }
```
Comment (asnare) [inline src/main/java/io/divolte/server/IncomingRequestProcessingPool.java]: Is this code in or out? :)

Comment (friso) [inline src/main/java/io/divolte/server/IncomingRequestProcessingPool.java]: Hmm... It is apparently untouched and the build is green, so I think it's
out.

File: src/main/java/io/divolte/server/ValidatedConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+ */
+@ParametersAreNonnullByDefault
+public final class ValidatedConfiguration {
+	private final static Logger logger = LoggerFactory.getLogger(ValidatedConfiguration.class);
+	
+    private final List<ConfigException> exceptions;
+    private final DivolteConfiguration divolteConfiguration;
+
+    /**
+	 * Creates an instance of a validated configuration. The underlying
```
Comment (asnare) [inline src/main/java/io/divolte/server/ValidatedConfiguration.java]: This files seems to contain a lot of tabs…

File: src/main/java/io/divolte/server/ValidatedConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+	 * Returns the validated configuration object tree. This is only returned
+	 * when no validation errors exist. The method throws
+	 * {@code IllegalStateException} otherwise.
+	 * 
+	 * @return The validated configuration.
+	 * @throws IllegalStateException
+	 *             When validation errors exist.
+	 */
+    public DivolteConfiguration configuration() {
+        if (exceptions.size() > 0) {
```
Comment (asnare) [inline src/main/java/io/divolte/server/ValidatedConfiguration.java]: .isEmpty()

(PMD should complain about this?)

File: src/main/java/io/divolte/server/ValidatedConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+        return exceptions;
+    }
+
+    /**
+	 * Returns false if validation errors exist, true otherwise.
+	 * 
+	 * @return false if validation errors exist, true otherwise.
+	 */
+    public boolean isValid() {
+        return exceptions.size() == 0;
```
Comment (asnare) [inline src/main/java/io/divolte/server/ValidatedConfiguration.java]: Ditto.

File: src/main/java/io/divolte/server/ValidatedConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+    @ParametersAreNullableByDefault
+    public final static class DivolteConfiguration {
+        public final ServerConfiguration server;
+        public final TrackingConfiguration tracking;
+        public final JavascriptConfiguration javascript;
+        public final IncomingRequestProcessorConfiguration incomingRequestProcessor;
+        public final KafkaFlusherConfiguration kafkaFlusher;
+        public final HdfsFlusherConfiguration hdfsFlusher;
+
+        public DivolteConfiguration(
```
Comment (asnare) [inline src/main/java/io/divolte/server/ValidatedConfiguration.java]: I think this constructor should be private, or we need to guard against null parameters.

File: src/main/java/io/divolte/server/ValidatedConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+                getOptionalOrAddException(      config::getString,      "divolte.tracking.schema_file",                 exceptions, config),
+                schemaMapping
+                );
+
+        final JavascriptConfiguration javascript = new JavascriptConfiguration(
+                getOrAddException(              config::getString,      "divolte.javascript.name",                      exceptions),
+                getOrAddException(              config::getBoolean,     "divolte.javascript.logging",                   exceptions),
+                getOrAddException(              config::getBoolean,     "divolte.javascript.debug",                     exceptions));
+
+        if (javascript.name != null && !javascript.name.matches("^[A-Za-z0-9_-]+\\.js$")) {
```
Comment (asnare) [inline src/main/java/io/divolte/server/ValidatedConfiguration.java]: I think this null-check is always true.

File: src/main/java/io/divolte/server/ValidatedConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+                final String publishDir) {
+            this.type = type;
+            this.syncDileAfterRecords = syncFileAfterRecords;
+            this.syncFileAfterDuration = syncFileAfterDuration;
+            this.workingDir = workingDir;
+            this.publishDir = publishDir;
+        }
+
+        public static enum Types {
+            SIMPLE_ROLLING_FILE, SESSION_BINNING;
```
Comment (asnare) [inline src/main/java/io/divolte/server/ValidatedConfiguration.java]: No trailing semicolon needed here.

File: src/main/java/io/divolte/server/ValidatedConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+
+    @ParametersAreNullableByDefault
+    public static class FileStrategyConfiguration {
+        public final Types type;
+        public final Integer syncDileAfterRecords;
+        public final Duration syncFileAfterDuration;
+        public final String workingDir;
+        public final String publishDir;
+
+        protected FileStrategyConfiguration (
```
Comment (asnare) [inline src/main/java/io/divolte/server/ValidatedConfiguration.java]: This constructor can also be private, I think.

File: src/main/java/io/divolte/server/ValidatedConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+    public static final class KafkaFlusherConfiguration {
+        public final Boolean enabled;
+        public final Integer threads;
+        public final Integer maxWriteQueue;
+        public final Duration maxEnqueueDelay;
+        public final String topic;
+        public final Properties producer;
+
+        private KafkaFlusherConfiguration(
+                final Boolean enabled, Integer threads,
```
Comment (asnare) [inline src/main/java/io/divolte/server/ValidatedConfiguration.java]: Indentation consistency.

File: src/main/java/io/divolte/server/ValidatedConfiguration.java (hunk 2 above)
Comment (friso) [inline src/main/java/io/divolte/server/ValidatedConfiguration.java]: Oops. Had to re-install my Eclipse and throw away my workspaces and everything to get it back into shape again. Will re-configure it to use spaces.

File: src/main/java/io/divolte/server/ValidatedConfiguration.java (hunk 5 above)
Comment (friso) [inline src/main/java/io/divolte/server/ValidatedConfiguration.java]: Yes, it needs to be private.

File: src/main/java/io/divolte/server/ValidatedConfiguration.java (hunk 6 above)
Comment (friso) [inline src/main/java/io/divolte/server/ValidatedConfiguration.java]: getOrAddException can return null.

File: src/main/java/io/divolte/server/ValidatedConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+        private HdfsConfiguration(Optional<String> uri, Short replication) {
+            this.uri = uri;
+            this.replication = replication;
+        }
+    }
+
+    @ParametersAreNullableByDefault
+    public static class FileStrategyConfiguration {
+        public final Types type;
+        public final Integer syncDileAfterRecords;
```
Comment (asnare) [inline src/main/java/io/divolte/server/ValidatedConfiguration.java]: Typo in name.

**Pull request merged**